src/edu/washington/escience/myria/operator/CountFilter.java

```diff
@@ -0,0 +1,214 @@
+import edu.washington.escience.myria.column.Column;
+import edu.washington.escience.myria.column.IntArrayColumn;
+import edu.washington.escience.myria.storage.MutableTupleBuffer;
+import edu.washington.escience.myria.storage.TupleBatch;
+import edu.washington.escience.myria.storage.TupleUtils;
+import edu.washington.escience.myria.util.HashUtils;
+
+/**
+ * Keeps distinct tuples with their counts and only emits a tuple at the first time when its count hits the threshold.
+ * */
+public final class CountFilter extends StreamingState {
+
+  /** Required for Java serialization. */
+  private static final long serialVersionUID = 1L;
+
+  /**
+   * The logger for this class.
+   * */
+  static final Logger LOGGER = LoggerFactory.getLogger(CountFilter.class);
+
+  /**
+   * Indices to unique tuples.
+   * */
+  private transient IntObjectHashMap<IntArrayList> uniqueTupleIndices;
+
+  /**
+   * The buffer for storing unique tuples.
+   * */
+  private transient MutableTupleBuffer uniqueTuples = null;
+  /**
+   * The count of each unique tuple.
+   * */
+  private transient MutableTupleBuffer tupleCounts = null;
+
+  /** threshold of the count. */
+  public final int threshold;
+
+  /** column indices of the key. */
+  private final int[] keyColIndices;
+
+  /**
+   * @param threshold threshold
+   * @param keyColIndices key column indices.
+   */
+  public CountFilter(final int threshold, final int[] keyColIndices) {
+    Preconditions.checkArgument(threshold >= 0, "threshold needs to be greater than or equal to 0");
+    this.threshold = threshold;
+    this.keyColIndices = Arrays.copyOf(keyColIndices, keyColIndices.length);
+  }
+
+  @Override
+  public void cleanup() {
+    uniqueTuples = null;
+    uniqueTupleIndices = null;
+    tupleCounts = null;
+  }
+
+  /**
+   * Do duplicate elimination for tb.
+   * 
+   * @param tb the TupleBatch for performing DupElim.
+   * @return the duplicate eliminated TB.
+   * */
+  protected TupleBatch countFilter(final TupleBatch tb) {
+    final int numTuples = tb.numTuples();
+    if (numTuples <= 0) {
+      return tb;
+    }
+    doCount.inputTB = tb;
+    final List<? extends Column<?>> columns = tb.getDataColumns();
+    final BitSet toRemove = new BitSet(numTuples);
+    for (int i = 0; i < numTuples; ++i) {
+      final int nextIndex = uniqueTuples.numTuples();
+      final int cntHashCode = HashUtils.hashSubRow(tb, keyColIndices, i);
+      IntArrayList tupleIndexList = uniqueTupleIndices.get(cntHashCode);
+      if (tupleIndexList == null) {
+        tupleIndexList = new IntArrayList();
+        uniqueTupleIndices.put(cntHashCode, tupleIndexList);
+      }
+      doCount.found = false;
+      doCount.meet = false;
+      doCount.sourceRow = i;
+      tupleIndexList.forEach(doCount);
+      if (!doCount.found) {
+        for (int j : keyColIndices) {
+          uniqueTuples.put(j, columns.get(j), i);
+        }
+        tupleCounts.put(0, new IntArrayColumn(new int[] { 1 }, 1), 0);
```

Comment:
Why can't you just use `putInt()` here and not construct this bogus singleton column?